### PR TITLE
MAGECLOUD-4053: Windows compatibility is broken

### PR DESCRIPTION
**Pull request by shiftedreality**
### Fixed Issues (if relevant)

1. https://magento2.atlassian.net/browse/MAGECLOUD-4053
2. https://github.com/magento/magento-cloud-docker/issues/69

### Manual testing scenarios

For Windows:

> Add `.env file` with
`WINDOWS_PWD=PATH_TO_PROJECT_FOLDER`

1. https://jira.corp.magento.com/browse/MAGECLOUD-66
2. https://jira.corp.magento.com/browse/MAGECLOUD-116
3. On Windows: https://jira.corp.magento.com/browse/MAGECLOUD-66
4. Using option `--no cron 1` https://jira.corp.magento.com/browse/MAGECLOUD-66

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages

### CODE REVIEW

File: config/services.xml [hunk 1]
```diff
@@ -7,13 +7,18 @@
...
         <prototype namespace="Magento\CloudDocker\" resource="../src/*" exclude="../src/{App,Test}"/>
-        <service id="Magento\CloudDocker\App\GenericException" autowire="false"/>
-        <service id="Magento\CloudDocker\App\ConfigurationMismatchException" autowire="false"/>
+        <service id="Magento\CloudDocker\Compose\BuilderFactory">
+            <argument key="$strategies" type="collection">
+                <argument key="developer">Magento\CloudDocker\Compose\DeveloperBuilder</argument>
+                <argument key="production">Magento\CloudDocker\Compose\ProductionBuilder</argument>
+                <argument key="functional">Magento\CloudDocker\Compose\FunctionalBuilder</argument>
+            </argument>
+        </service>
```
Comment (bbatsche) [inline config/services.xml:16]: Why the change from using the map constant to passing in through configuration? I think either way is fine, I'm just curious about the benefit?

File: src/Compose/FunctionalBuilder.php [hunk 2]
```diff
@@ -7,28 +7,66 @@
...
+        Reader $reader
+    ) {
+        $this->fileList = $fileList;
+
+        parent::__construct(
+            $serviceFactory,
+            $serviceConfig,
+            $fileList,
+            $directoryList,
+            $converter, $phpExtension,
```
Comment (bbatsche) [inline src/Compose/FunctionalBuilder.php]: Need a new line here

File: src/Compose/FunctionalBuilder.php [hunk 3]
```diff
@@ -7,28 +7,66 @@
...
+    /**
+     * @param ServiceFactory $serviceFactory
+     * @param Config $serviceConfig
+     * @param FileList $fileList
+     * @param DirectoryList $directoryList
+     * @param Converter $converter
+     * @param ExtensionResolver $phpExtension
+     * @param Reader $reader
+     */
+    public function __construct(
```
Comment (bbatsche) [inline src/Compose/FunctionalBuilder.php:42]: I'm not sure why you're overriding the `__construct()` method here, it looks like it takes all the same arguments as its parent?

File: src/Service/ServiceFactory.php [hunk 4]
```diff
@@ -17,31 +17,47 @@
...
-    const CONFIG = [
+    /**
+     * @var array
+     */
+    private static $config = [
         self::SERVICE_CLI => [
-            'image' => 'magento/magento-cloud-docker-php:%s-cli-%s'
+            'image' => 'magento/magento-cloud-docker-php:%s-cli-%s',
+            'config' => [
+                'extends' => 'generic'
```
Comment (bbatsche) [inline src/Service/ServiceFactory.php]: Should these all be `'extends' => static::SERVICE_GENERIC`?

**Review by bbatsche — CHANGES_REQUESTED** [review]

File: config/services.xml (hunk 1 above)
Comment (shiftedreality) [reply config/services.xml:16]: Technically Functional builder must be within ECE-Tools repo. Adding Symfony DI later will give as a possibility to inject a line with dependency

File: src/Compose/FunctionalBuilder.php (hunk 2 above)
Comment (shiftedreality) [reply src/Compose/FunctionalBuilder.php]: Fixed

File: src/Compose/FunctionalBuilder.php (hunk 3 above)
Comment (shiftedreality) [reply src/Compose/FunctionalBuilder.php:42]: I need one depnendency `$fileList`

File: src/Service/ServiceFactory.php (hunk 4 above)
Comment (shiftedreality) [reply src/Service/ServiceFactory.php]: Fixed

File: src/Compose/FunctionalBuilder.php (hunk 3 above)
Comment (bbatsche) [reply src/Compose/FunctionalBuilder.php:42]: Ok, but `FileList` is already a dependency of `ProductionBuilder`, you just changed it from `protected` to `private` in this very same PR. Leave it `protected` and there's no need to override `__construct()` here.

**Review by bbatsche — CHANGES_REQUESTED** [review]

**Comment by shiftedreality** [comment]
We should avoid protected modifier, it's used only in few methods as it's too tricky to do this without inheritance

File: src/Compose/FunctionalBuilder.php (hunk 3 above)
Comment (shiftedreality) [reply src/Compose/FunctionalBuilder.php:42]: We should avoid protected modifier, it's used only in few methods as it's too tricky to do this without inheritance

**Review by bbatsche — APPROVED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Review by BaDos — APPROVED** [review]

**Comment by oshmyheliuk** [comment]
QA on the Windows environment is approved.
For me next format `WINDOWS_PWD=/C/path/to/magento/cloud/` works fine.

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**